text/0000-const-ub.md

```diff
@@ -0,0 +1,116 @@
+So either the code is stored twice (optimized and unoptimized), or optimizations can only happen after all CTFE results have been computed.
+[Experiments in rustc](https://perf.rust-lang.org/compare.html?start=35debd4c111610317346f46d791f32551d449bd8&end=3dbdd3b981f75f965ac04452739653a3d47ff0ed) showed a severe performance impact on CTFE stress-tests, but no impact on real code except for a slowdown of "incr-unchanged" (which are rather fast so small changes lead to large percentages).
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The most obvious alternative is to say that UB during CTFE will definitely be detected.
+However, that is expensive and might even be impossible.
+Even Miri does not currently detect all UB, and Miri is already performing many additional checks that would significantly slow down CTFE.
+Furthermore, implementing these checks requires a more precise understanding of UB than we currently have; basically, this would block having any potentially-UB operations at const-time on having a spec for Rust that precisely describes their UB in a checkable way.
+In particular, this would mean we need to decide on an aliasing model before permitting raw pointers in CTFE.
+
+To avoid the need for keeping the unoptimized sources of `const fn` around, we could weaken the requirement for detecting UB and instead say that UB might cause arbitrary evaluation results.
+Under the assumption that unsound code is not subject to the usual stability guarantees, this is an option we can still move to in the future, should it turn out that the proposal made in this RFC is too expensive.
+
+Another extreme alternative would be to say that UB during CTFE may have arbitrary effects in the host compiler, including host-level UB.
+Basically this would mean that CTFE would be allowed to "leave its sandbox".
+This would allow JIT'ing CTFE and running the resulting code unchecked.
+While compiling untrusted code should only be done with care (including additional sandboxing), this seems like an unnecessary extra footgun.
+
+# Prior art
+[prior-art]: #prior-art
+
+C++ requires compilers to detect UB in `constexpr`.
+However, the fragment of C++ that is available to `constexpr` excludes pointer casts, pointer arithmetic (beyond array bounds), and union-based type punning, which makes such checks not very complicated and avoids most of the poorly specified parts of UB.
+The corresponding type-punning-free fragment of Rust (no raw pointers, no `union`, no `transmute`) can only cause UB that is defined to be definitely detected during CTFE.
+In that sense, Rust achieves feature parity with C++ in terms of UB detection during CTFE.
+(Indeed, this was the prime motivation for making such strict UB detection requirements in the first place.)
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Currently none.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
```

Comment:
I think it might be useful to add a `--fewer-ctfe-ub-checks` flag at some point in the future. This flag would allow running CTFE on optimized MIR, but would relax the guarantees about which kinds of UB would be detected. This would only change the behavior of incorrect code, and could speed up compilation for users who have audited their unsafe code.


This would need additional design work to be properly integrated with cargo (something per-crate would be better than `RUSTFLAGS="--fewer-ctfe-ub-checks"`), but I think it would be good for the RFC to leave the possibility open.

Reply:
Thanks, I added that.